packages/core-data/src/actions.js

```diff
@@ -209,29 +214,73 @@ export function* redo() {
  * @param {string} kind    Kind of the received entity.
  * @param {string} name    Name of the received entity.
  * @param {Object} record  Record to be saved.
+ * @param {Object} options Saving options.
  */
-export function* saveEntityRecord( kind, name, record ) {
+export function* saveEntityRecord(
+	kind,
+	name,
+	record,
+	{ isAutosave = false } = { isAutosave: false }
+) {
 	const entities = yield getKindEntities( kind );
 	const entity = find( entities, { kind, name } );
 	if ( ! entity ) {
 		return;
 	}
-	const key = entity.key || DEFAULT_ENTITY_KEY;
-	const recordId = record[ key ];
+	const entityIdKey = entity.key || DEFAULT_ENTITY_KEY;
+	const recordId = record[ entityIdKey ];
 
-	yield { type: 'SAVE_ENTITY_RECORD_START', kind, name, recordId };
+	yield { type: 'SAVE_ENTITY_RECORD_START', kind, name, recordId, isAutosave };
 	let error;
 	try {
-		const updatedRecord = yield apiFetch( {
-			path: `${ entity.baseURL }${ recordId ? '/' + recordId : '' }`,
-			method: recordId ? 'PUT' : 'POST',
-			data: record,
-		} );
-		yield receiveEntityRecords( kind, name, updatedRecord, undefined, true );
+		const path = `${ entity.baseURL }${ recordId ? '/' + recordId : '' }`;
+		if ( isAutosave ) {
+			const persistedRecord = yield select(
+				'getEntityRecord',
+				kind,
+				name,
+				recordId
+			);
+			const currentUser = yield select( 'getCurrentUser' );
+			const currentUserId = currentUser ? currentUser.id : undefined;
+			const autosavePost = yield select(
+				'getAutosave',
+				persistedRecord.type,
+				persistedRecord.id,
+				currentUserId
+			);
+			let data = { ...persistedRecord, ...autosavePost, ...record };
```

Comment:
I'm having trouble understanding which data we're sending.  Maybe a comment could help. Like do we send just specific record properties, everything, why are we merging these three records, and why that order.

Reply:
I've added this comment:

>// Autosaves need all expected fields to be present.
			// So we fallback to the previous autosave and then
			// to the actual persisted entity if the edits don't
			// have a value.

Reply:
I guess I'm wondering why we need the three steps merge. If I'm undersanding properly, we don't really need `autosavePost` here since the edits will most likely contain these as well right?

Reply:
Not always.

It has 3 items because we have 2 fallbacks. First, the previous autosave, and then the persisted values.

Reply:
Really? The UI only shows the `edits` and not the autosaved content so I'm having trouble thinking of use-cases where autosaves have more data than edits.

Reply:
Imagine you only edit the title and that autosaves. You would have an autosave with no post content.

Reply:
Autosaving is not patching an existing post, it creates a new post.

Reply:
yes, but you'd have the `content` in the persisted post

Reply:
Yes, but you still need it in the autosave.

Reply:
`persistedRecord` above would contain `content` If I'm not wrong, so it will be in the autosave? Sorry If I'm missing something obvious.

Reply:
Yes, that's why it's a fallback. But the previous autosave is a fallback with higher precedence, because it might have the explicit edits passed in a previous autosave that the persisted record does not have yet.

Reply:
I guess what I'm saying is that I don't see any usecase where this fallback is necessary. Since persisted values will be in the persisted post and edits in the `record`. If  we look at the function itself `saveEntityRecord` with `autosave` true. Nothing indicates that it's going to use a previous autosave to save the edits. It's just saying: create an autosave for this post with these changes which suggests to me that we shouldn't fallback to the previous autosave. Also, we'd avoid two selectors calls `getAutosave`  and `getCurrentUser`.

Reply:
- Creates new post and saves it with `title: Hello` and `content: World`.
- Autosaves content edit which uses title fallback from persisted post. Autosave now contains `title: Hello` and `content: My autosaved content.`.
- Autosaves title edit. If we drop the first fallback like you are suggesting the new autosave would have `title: My Autosaved Title` and `content: World`. Because we do have the fallback, it will have `title: My Autosaved Title` and `content: My autosaved content.`.

Reply:
>Nothing indicates that it's going to use a previous autosave to save the edits.

That's just an implementation detail required because of how the autosaves API works.

Reply:
> Autosaves content edit which uses title fallback from persisted post. Autosave now contains title: Hello and content: My autosaved content..

Mmm, what happens in the UI when we open the post at that point? Do we load both the edits and the autosave in UI or do we load the persisted post with a message saying that there's an autosave?

If it's the former, I agree with you, we should merge with the autosave, if it's the latter, we shouldn't.

Reply:
When you open the post at that point there wouldn't be any edits.

It just shows the message that there is an autosave. We use this fallback mechanism in the editor store right now.

You are still risking content lost. What if a different block that saves to the same entity triggers an autosave with `title` and that clears the `content` autosaved by another block.

Reply:
I'd say that's the expected behavior because the autosave should save only what's in the screen? not something I'm not seeing anywhere unless I click the button to "recover the autosave" which I assume will copy the autosave to the "edits".

Reply:
These blocks could be rendered at the same time or on different template posts. At that point you would not be able to recover the previous autosave with the current workflow.

Reply:
you would but that's an explicit action. `saveEntityRecord( getAutosave() )`

Reply:
That would not work, you would need to merge the properties like in this PR.

Why complicate things when this will be the only usage of the API. When would you want to clear autosaved properties without changes?

Reply:
How does it work today in the editor screen without the refactoring? Say I have an autosave but I ignore it and I continue editing my post, another autosave will trigger right without the updates from the previous autosave?

How would you do that in this new approach? 

To be honest, I don't want to complicate anything.  I actually think I'm simplifying things. Anyway, I won't block that PR for this, I feel we can tweak this later if needed but still cuirious about the answer to the question above.

Reply:
>>We use this fallback mechanism in the editor store right now.

It works like this right now.

>Say I have an autosave but I ignore it and I continue editing my post, another autosave will trigger right without the updates from the previous autosave?

It will have the updates from the previous autosave.

https://github.com/WordPress/gutenberg/blob/92918453c6c484a7f6cd8cb218c57456cfc2ccee/packages/editor/src/store/actions.js#L524-L528

Reply:
> It works like this right now.

Ok let's keep it but it feels like a bug to me. cc @adamsilverstein if you know whether an autosave should be created based on the previous autosave or based on the things that are shown in the UI (previous saved state + edits)